File: solutions/0720. Longest Word in Dictionary/0720.java
class Solution {
  public String longestWord(String[] words) {
    for (final String word : words)
      insert(word);

    return dfs(root);
  }

  private class TrieNode {
    private TrieNode[] children = new TrieNode[26];
    private String word;
  }

  private TrieNode root = new TrieNode();

  private void insert(final String word) {
    TrieNode node = root;
    for (char c : word.toCharArray()) {
      int index = c - 'a';
      if (node.children[index] == null)
        node.children[index] = new TrieNode();
      node = node.children[index];
    }
    node.word = word;
  }

  private String dfs(TrieNode node) {
    String ans = node.word == null ? "" : node.word;

    for (TrieNode child : node.children)
      if (child != null && child.word != null && child.word.length() > 0) {
        String childWord = dfs(child);
        if (childWord.length() > ans.length())
          ans = childWord;
      }

    return ans;
  }
}

File: solutions/0720. Longest Word in Dictionary/0720.py
class Solution:
  def longestWord(self, words: List[str]) -> str:
    def dfs(node: dict) -> str:
      ans = node['word'] if 'word' in node else ''

      for child in node:
        if 'word' in node[child] and len(node[child]['word']) > 0:
          childWord = dfs(node[child])
          if len(childWord) > len(ans) or (len(childWord) == len(ans) and childWord < ans):
            ans = childWord

      return ans

    self.root = {}

    for word in words:
      node = self.root
      for c in word:
        if c not in node:
          node[c] = {}
        node = node[c]
      node['word'] = word

    return dfs(self.root)
